[noetic][pluginlib] declare specific boost dependencies

mikaelarguedas opened the pull request:
May be worth targeting to a noetic-devel to not impact released packages on other distributions

related to https://github.com/ros/ros_comm/pull/1871

mikaelarguedas committed:
[pluginlib] declare specific boost dependencies

Signed-off-by: Mikael Arguedas <mikael.arguedas@gmail.com>

caioaamaral commented on pluginlib/package.xml:
```diff
@@ -22,7 +22,8 @@
 
-  <depend>boost</depend>
+  <depend>libboost-filesystem-dev</depend>
+  <depend>libboost-system-dev</depend>
```
Wouldn't be better having those inside a <build_depend> instead?

As state here: https://docs.ros.org/indigo/api/catkin/html/howto/format2/system_library_dependencies.html

> It is not recommended for system dependencies, because it forces your package’s binary installation to depend on the “development” package, which is not generally necessary or desirable

mikaelarguedas replied:
Thanks for the feedback

This PR has been mostly automated as a first step in an effort to get rid of `libboost-all-dev` in all of `ros_core`, and did not audit each package closely.
Currently pluginlib is using `<depend>boost</depend>` that pulls `libboost-all-dev` / ~800MB of development packages.
This change is an incremental improvement as after this change it will pull "only" 200.

---

Looking closer at this package:
Pluginlib is a header-only library, so it *has to* export the dependencies downstream users need install to be able to link against pluginlib. So it cannot be only a `build_depend`.

It can however be changed to be a simple `build_export_depend`. The same would apply to all `<depend>` listed in this package.

mikaelarguedas committed:
convert depend in build_export_depend as pluginlib is header only

actually keep them as build_depend and build_export_depend as they are "find_package(REQUIRED) in the CMakeLists of this package"

remove libboost-system as it doesnt seem to be imported/used anywhere
Signed-off-by: Mikael Arguedas <mikael.arguedas@gmail.com>

sloretz commented:
@ros-pull-request-builder retest this please

sloretz commented:
@ros-pull-request-builder retest this please

sloretz commented:
CI looks ok, just a CMP0048 warning which could be fix in a different PR: http://wiki.ros.org/noetic/Migration#Increase_required_CMake_version_to_avoid_author_warning

sloretz commented on pluginlib/package.xml:
```diff
@@ -21,12 +21,17 @@
-  <depend>rosconsole</depend>
-  <depend version_gte="1.11.1">roslib</depend>
-  <depend>tinyxml2</depend>
+  <build_export_depend>libboost-filesystem-dev</build_export_depend>
```
Does this still bring in `boost/algorithm`, `boost/bind`, `boost/foreach`, and `boost/shared_ptr`?

https://github.com/ros/pluginlib/blob/dfd49dad457d81f4fc5984a198f4b2abc39ba7c6/pluginlib/include/pluginlib/class_loader.hpp#L37

https://github.com/ros/pluginlib/blob/dfd49dad457d81f4fc5984a198f4b2abc39ba7c6/pluginlib/include/pluginlib/class_loader_imp.hpp#L50-L53

https://github.com/ros/pluginlib/blob/dfd49dad457d81f4fc5984a198f4b2abc39ba7c6/pluginlib/test/utest.cpp#L103

mikaelarguedas replied:
Good question,

Any `libboost-LIBNAME-dev` package depends on `libboost-dev` that brings in ALL boost headers.

Then if we want to be sure these headers are still brought in the day the filesystem dependency is removed from pluginlib, and no other dependency brings boost headers in (very unlikely), we need to add a dependency on `libboost-dev` in this package.

sloretz commented:
Opened #173 to fix Noetic pluginlib PR jobs

nuclearsandwich commented:
@ros-pull-request-builder retest this please

mikaelarguedas committed:
add libboost-dev to allow filesystem removal in the future

Signed-off-by: Mikael Arguedas <mikael.arguedas@gmail.com>

sloretz commented:
@nuclearsandwich friendly ping. Looks like both Melodic and Noetic PR jobs passed :tada:

sloretz commented:
@nuclearsandwich friendly ping

sloretz committed:
Merge branch 'noetic-devel' into restrict-boost-deps

sloretz commented:
Results of a Noetic prerelease test

```
mkdir -p /tmp/prerelease_job
cd /tmp/prerelease_job
generate_prerelease_script.py \
  https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/production/index.yaml \
  noetic default ubuntu focal amd64 \
  --custom-repo \
    pluginlib__custom-1:git:https://github.com/mikaelarguedas/pluginlib.git:restrict-boost-deps \
  --level 7 \
  --output-dir ./ \
  --exclude-pkg capabilities fkie_message_filters
```

Plus edited overlay to include https://github.com/at-wat/neonavigation/pull/497

```
ws_overlay/test_results/controller_manager_tests/rostest-test_controller_manager_scripts.xml: 2 tests, 1 errors, 0 failures, 0 skipped
ws_overlay/test_results/geodesy/nosetests-tests.test_wu_point.py.xml: 5 tests, 2 errors, 0 failures, 0 skipped
ws_overlay/test_results/rosmsg/nosetests-test.xml: 31 tests, 3 errors, 4 failures, 0 skipped
ws_overlay/test_results/stage_ros/rostest-test_intensity_test.xml: 1 tests, 0 errors, 1 failures, 0 skipped
ws_overlay/test_results/ypspur_ros/rosunit-test_joint_trajectory.xml: 2 tests, 0 errors, 2 failures, 0 skipped
Summary: 2394 tests, 6 errors, 7 failures, 0 skipped
```


* controller_manager_tests/rostest-test_controller_manager_scripts.xml: 2 tests, 1 errors, 0 failures, 0 skipped
  * [Not failing in dev job](http://build.ros.org/view/Ndev/job/Ndev__ros_control__ubuntu_focal_amd64/7/testReport/rostest.runner/RosTest/testcontroller_manager_scripts_test/); needs investigation
* geodesy/nosetests-tests.test_wu_point.py.xml: 5 tests, 2 errors, 0 failures, 0 skipped
  * Python 3 incompatibility; missing xrange
* rosmsg/nosetests-test.xml: 31 tests, 3 errors, 4 failures, 0 skipped
  * resource not found `test_rosmaster`, seems like `rosmsg` tests are expecting source repo layout?
* stage_ros/rostest-test_intensity_test.xml: 1 tests, 0 errors, 1 failures, 0 skipped
  * `[intensity_check] did not generate test results`, [also happening in dev job](http://build.ros.org/view/Ndev/job/Ndev__stage_ros__ubuntu_focal_amd64/1/testReport/junit/rostest.runner/RosTest/testintensity_check/)
* ypspur_ros/rosunit-test_joint_trajectory.xml: 2 tests, 0 errors, 2 failures, 0 skipped
  * `Value of: static_cast<bool>(joint_states)` is `false`, [also happening in dev job](http://build.ros.org/view/Ndev/job/Ndev__ypspur_ros__ubuntu_focal_amd64/2/testReport/(root)/JointTrajectory/CommandValidation/)


So looking pretty good. The controller_manager_tests failure is the only one I want to check out.

sloretz commented:
I've run the controller_manager_tests a couple times by re-running `./prerelease_build_overlay.sh` and haven't seen the failures again. I'm going to assume it's a flaky test, and none of the other test results seem to be the result of this PR.

@nuclearsandwich I'm going to use the permission you gave me at the start of the day to merge and release.
